Proposal: edits to RELEASE_GUIDE.md (#4104)
purescript/purescript: RELEASE_GUIDE.md

# Release Guide (for maintainers)

## Prerequisites

- You will need a [Hackage](https://hackage.haskell.org/) account that has been invited to be a maintainer of the `purescript` package on Hackage. If you don't have one, create one and ask to be invited as a maintainer.
- You will need an [NPM](https://www.npmjs.com/) account that has been invited to be a maintainer of the `purescript` package on NPM. If you don't have one, create one and ask to be invited as a maintainer.

## Before making a release

- Check that there are no unintended breaking changes by compiling [the latest package set](https://github.com/purescript/package-sets/releases/latest)

```bash
stack build
mkdir wPackageSet
pushd wPackageSet
spago init
spago upgrade-set
# install all packages in the set

> spago install $(spago ls packages | cut -f 1 -d ' ' | tr '\n' ' ')

Reviewer: We really need a `spago install --whole-set` 😄

Reviewer 2: Is this just `spago verify-set`?

Author: I think it would replace the `spago install` and `spago build` command. However, I am not sure whether it also builds docs. 

@f-f, could you clarify?

Reviewer: The question "can I install the entirety of the set?" seems to come up often, and I don't have a good answer for that right now (it will improve once we get the new config format for Spago).

In this thread however, I think Harry is right: installing the whole set and then building it is _basically_ `verify-set` (which does some more stuff, so it's going to be slightly slower than just installing everything and building it, but it's fine)

stack exec purs compile $(spago sources | tr '\n' ' ')

> stack exec purs docs $(spago sources)

Author: Note: these two `stack exec` commands don't work. I'm not sure how to pass the source globs correctly to the resulting `purs` binary.

Reviewer: If you run `stack exec -- bash` then you'll be dropped into a shell, so maybe that'd be easier?

Reviewer 2: Yeah, I think `stack exec bash` followed by `spago build` and `spago docs` is probably easiest.

Author: I tried `stack exec bash` and it didn't drop into a shell. Found [this SO answer](https://stackoverflow.com/a/48014385) and think we should use this:
```bash
stack exec bash <<EOF
spago build
spago docs
EOF
```

# Release Guide (for maintainers)

## Prerequisites

- You will need a [Hackage](https://hackage.haskell.org/) account that has been invited to be a maintainer of the `purescript` package on Hackage. If you don't have one, create one and ask to be invited as a maintainer.
- You will need an [NPM](https://www.npmjs.com/) account that has been invited to be a maintainer of the `purescript` package on NPM. If you don't have one, create one and ask to be invited as a maintainer.
- You need `spago` installed.
- You need to be logged into NPM (i.e. running `npm whoami` should print your NPM account's username)

## Before making a release

- Check that there are no unintended breaking changes by compiling [the latest package set](https://github.com/purescript/package-sets/releases/latest)

```bash
stack build
mkdir wPackageSet
pushd wPackageSet
spago init
spago upgrade-set
# install all packages in the set
spago install $(spago ls packages | cut -f 1 -d ' ' | tr '\n' ' ')

# Verify that code compiles and docs are properly created

> exec stack exec bash <<EOF

Reviewer 2: You probably don't want the first `exec` here. It will make your terminal/shell exit after the command is complete.
```suggestion
stack exec bash <<EOF
```

Author: Whoops!

Author: This was fixed in latest commits.

spago build
spago docs
EOF
popd
# rm -rf wPackageSet
```

- Check that INSTALL.md is up-to-date. For example:
    - Does the GHC version correspond to the one we're currently on?
    - Is the list of supported OSes (based on GHC version) correct? If not, check the GHC website for the list.
    - Are there any unofficial installation methods that should be dropped because they are no longer maintained?
    - Are the instructions for building from source currently working?
    - Anything else that needs to be changed?

- Additionally, if there are any breaking changes, there are number of downstream
projects whom we should probably at least notify. See below subsections:

### Libraries

Are there breaking changes to the language? Or alternatively, are there
language changes which require breaking changes in the relevant libraries to
make use of? If so, make a "meta" issue to track which libraries need to be changed and then update those libraries:

> - core libs: open a "meta" issue in [`purescript/purescript` issue](https://github.com/purescript/purescript/issues)). For example, [#3942](https://github.com/purescript/purescript/issues/3942)
> - contrib libs: open a "meta" issue in [`purescript-contrib/governance` issue](https://github.com/purescript-contrib/governance/issues). For example, [#35](https://github.com/purescript-contrib/governance/issues/35)
> - node libs: open a "meta" issue in [`purescript-contrib/governance` issue](https://github.com/purescript-contrib/governance/issues). For example, [#35](https://github.com/purescript-contrib/governance/issues/35)
> - web libs: open a "meta" issue in [`purescript-contrib/governance` issue](https://github.com/purescript-contrib/governance/issues). For example, [#35](https://github.com/purescript-contrib/governance/issues/35)

Reviewer 2: Typo here? "Open a meta issue in purescript/purescript issue" doesn't make sense to me. I'm also not sure we need to repeat the whole "open a "meta" issue" spiel for every item in this list.

```suggestion
- core libraries
- contrib libraries
- node libraries
- web libraries

It may be helpful to open a "meta" issue in the relevant repository to track progress for each group of libraries which need updating. For example, [purescript/purescript#3942](https://github.com/purescript/purescript/issues/3942).
```

## Making a release

> A release is made by updating the versions within the repo via a PR. Once the PR is merged, a GitHub release is drafted, which triggers a CI build that should upload prebuilt binaries into that GitHub release page. (If it doesn't, you'll upload those manually).

Reviewer 2: Can you avoid the passive voice for actions that the maintainer needs to take please? I think it is making it unclear which steps are performed automatically and which ones need to be done by the person making the release. Perhaps something like: "To make a release, a maintainer should create a PR updating versions in the repo, and once that is merged, they can create a GitHub release, which triggers a CI build ..."

After that, the respective packages on Hackage and NPM will be uploaded/updated.

### Set up a local branch for the PR

```bash
cd purescript
git fetch upstream
git checkout upstream/master
git switch -c prepareNextRelease
```

### Make last changes and submit the PR

```bash
# 1. Regenerate the LICENSE file and commit it
# See `./license-generator/` for details)
make license-generator
git add LICENSE
git commit -m "Regenerated LICENSE file"

# 2. Edit the changelog
  # 2a. Create a new "Unreleased" section
  # 2b. Rename the "Unreleased" section to
  #   the version you will be releasing
  # 2c. Edit the new version's content

>   #   until satisfied (e.g. grammer,

Reviewer 2: ```suggestion
  #   until satisfied (e.g. grammar,
```

  #   spelling, clarity, etc.)

EDITOR="nano" # or your preferred editor
$EDITOR CHANGELOG.md
git add CHANGELOG.md

> git commit -m "Added new Unreleased section; finalized new version's section"

Reviewer 2: I think "Finalizing CHANGELOG for {version}" might be a slightly clearer commit message, if we are going to suggest one here.

  # 3a. If it has changed since the
  #   last release, bump the version
  #   The new version should be based on the PVP,
  #   according to what changed since the previous release, #   not on the actual compiler version.
  #   See https://pvp.haskell.org/
$EDITOR lib/purescript-cst/purescript-cst.cabal

  # 3b. Bump the version
  # 3c. If it changed, update the `purescript-cst`
  #   version to match the new one
$EDITOR purescript.cabal

  # 3d. If making a pre-release,
  # change the `prerelease` string.
  # Otherwise, it should be "" for all
  # other releases.
$EDITOR app/Version.hs

  # 3e. Update the table
$EDITOR lib/purescript-cst/README.md

  # 3f. Change the "version" field
  # 3g. In the `postinstall` script,
  #   change the version.
$EDITOR npm-package/package.json

git add \
  lib/purescript-cst/purescript-cst.cabal \
  lib/purescript-cst/README.md \
  purescript.cabal \
  app/Version.hs \
  npm-package/package.json
git commit -m "Bumped versions for upcoming release"

# 4. Create the PR either manually or
#   via the `gh` CLI tool
gh pr create
# a. Make the `base` repo `purescript`
# b. Push your branch to your fork
# c. Continue on the web

> # 5. Push any other changes to the PR

Reviewer 2: I am not sure that the changes to this section are an improvement. They add quite a bit of noise, and explain how to do basic tasks like creating git branches, opening files in  their editor, or submitting PRs, but we should assume that a maintainer creating a release knows how to do these things. I think the extra noise is obscuring the important information here, which is which parts of which files need updating.

Author: Yeah... I have to agree with you here. It also doesn't account for people who are using editors that can easily navigate to such files.

#   until it gets merged.
```

### Make the GitHub Release

Once the above PR has been merged, create a release from the releases tab in GitHub, copying and pasting the version's corresponding section in the CHANGELOG.md file into the release as its release notes. This will also create a tag, which will kick off a CI build. The CI build will upload prebuilt compiler binaries to the release on GitHub when it completes.

If the GitHub-release-triggered CI build fails, the binaries will need to be built locally and manually uploaded to the release on GitHub.

> INSTRUCTIONS TO BE WRITTEN.

Author: Would this be 
```bash
# if Linux
CI_RELEASE="true" ./bundle/build.sh "Linux"
# if Mac
CI_RELEASE="true" ./bundle/build.sh "macOS"
# if Windows
CI_RELEASE="true" ./bundle/build.sh "Windows"
```

Reviewer 2: It would be along those lines, but you'll also need to make sure you've built a release version of the compiler first as well (so that `purs --version` doesn't say it's a development build). In the name of saving effort, since this scenario might not even happen, I'm tempted to say let's not bother writing instructions. Since we are the maintainers of the CI workflow/release script, we should be able to figure out how to run it if need be.

- JSON produced by `purs publish`
  - this might affect Pursuit

Collaborate with other maintainers to identify what work needs to be done to resolve any issue.

## Making a release

> To make a release, a maintainer should create a PR updating versions in the repo, and once that is merged, they can create a GitHub release, which triggers a CI build that should upload prebuilt binaries into that GitHub release page. (If it doesn't, you'll upload those manually).

Reviewer 2: You've mentioned this both at the start and the end of this section. I don't think it's necessary to mention it twice; can we remove one?

After that, the maintainer will upload the respective packages to Hackage and NPM.

### Make last changes and submit the PR

> 1. Checkout a new branch based on current `master`

Reviewer 2: Let's get rid of this too

```bash
git fetch upstream
git checkout upstream/master
git switch -c prepareNextRelease
```

2. Regenerate the LICENSE file and commit it. See `./license-generator/` for details.

```bash
make license-generator
git add LICENSE
git commit -m "Regenerated LICENSE file"
```

3. Edit the CHANGELOG.md file
    - Create a new "Unreleased" section
    - Rename the old "Unreleased" section to the version you will be releasing
    - Edit the new version's content until satisfied (e.g. grammar, spelling, clarity, etc.)

```bash
git add CHANGELOG.md
git commit -m "Finalizing CHANGELOG for {version}"
```

4. Make a commit bumping versions. The following should be updated:

- If `purescript-cst` has changed at all since the last release:

    - In `lib/purescript-cst/purescript-cst.cabal`, update the `version` field
      (note that the new version should be based on the PVP, according to what
      changed since the previous release, and not on the actual compiler
      version)

    - In `purescript.cabal`, update the version bounds for `purescript-cst`

  - In `lib/purescript-cst/README.md`, update the versions table to include a new entry for the new PureScript release

>   - In `purescript.cabal`, update the `version` field

Reviewer 2: This item and all of the other items following it need to be dedented so they aren't sub-items of the "If purescript-cst has changed" part, because all of these things need to happen every time, regardless of whether purescript-cst has changed.

1. From the [releases](https://github.com/purescript/purescript/releases) tab in GitHub, click the "Draft a new release" button.
2. In the "Tag version" and "Release title" fields, type in the version of the next release (e.g. "v0.14.2").
3. In the release notes, copy and paste the version's corresponding section in the CHANGELOG.md file.
4. Click the "Publish Release" button

This will create the tag defined above, which will kick off a CI build. When it completes, the CI build will upload prebuilt compiler binaries to the release on GitHub. You can see the CI build's progres via the [Actions](https://github.com/purescript/purescript/actions) tab under a workflow whose name matches the tag you defined previously.

If the GitHub-release-triggered CI build fails, the binaries will need to be built locally and manually uploaded to the release on GitHub. In the unlikely event this occurs, coordinate with other maintainers on how to resolve the issue.

### Publishing the packages to their respective registries

#### Hackage Uploads

```bash
cd purescript

# 1. Upload the `purescript-cst` package
pushd purescript/lib/purescript-cst

> stack upload

Reviewer 2: This doesn't work without the `.`
```suggestion
stack upload .
```

popd

# 2. Upload the `purescript` package`

> stack upload

Reviewer 2: ```suggestion
stack upload .
```

> # 3. Verify that the two (i.e. `purescript-cst` and `purescript`) can be installed from Hackage.
> 
> mkdir zHackageUpload
> pushd zHackageUpload
> cabal install purescript
> popd
> # rm -rf zHackageUpload

Reviewer 2: If we're using cabal, then there's no need to do this in a separate directory, because `cabal install xyz` installs the package globally and doesn't do anything with the current working directory. But I think we should avoid adding separate instructions here anyway, because we should be using the same instructions as we provide in INSTALL.md for installing the compiler from source.

```suggestion
# 3. Verify that `purescript-cst` and `purescript` can be installed from Hackage by following the instructions in INSTALL.md for building the compiler from source.
```
